[CL] Surface Sorption Process

renchao-lu opened the pull request:
This pull request is motivated for surface sorption/desorption simulation. For this purpose, we further extend the existing ChemistryLib. The newly added utilities are summarized as follows:
- Add keyword `SURFACE` for surface sorption/desorption simulation.
- Add keyword `KNOBS` for specifying numerical settings used in speciation calculation.
- Add keyword `USER_PUNCH` for output user-defined secondary variables in time-series `vtu` files.
- Add keyword `DUMP` for acquisition of aqueous solutions of last time step.
- Execute equilibrium calculation with regards to chemical systems in the initial phase.
- Optimized current file-based work flow.

1. [x] Feature description was added to the [changelog](https://github.com/ufz/ogs/wiki/Release-notes-6.2.2)
2. [x] Tests covering your feature were added?
3. [ ] Any new feature or behavior change was documented?

endJunction commented on Tests/Data/Parabolic/ComponentTransport/ReactiveTransport/SurfaceSorption/ESTRALdata_20120312_FINAL_GEH.dat:
```diff
@@ -0,0 +1,3547 @@
+
+# 			ESTRAL Thermodynamic Data Base
+#
+# Project: ESTRAL (FKZ 02 E 10528)              Developed 22-Jul-2009
```
This file needs a clear origin and a license.

montoyav replied:
@renchao-lu @endJunction I take of asking the group who provided us the database how they want we make reference to it. 
General comment: In my opinion, chemical thermodynamic databases used by OGS to perform "reactive processes" should be more clearly documented,  "visible". @endJunction maybe we speak about this and see what can be the best solution

joboog commented on ChemistryLib/PhreeqcIOData/Surface.h:
```diff
@@ -0,0 +1,46 @@
+struct SurfaceSite
+{
+    SurfaceSite(std::string&& mineral_,
+                double const site_density_,
```
please add a variable for the flag "-sites".
In Phreeqc you can specifiy a surface site by giving the *-site_density* in n/m2 or by giving the absolute amount of surface sites by *-sites*.

renchao-lu replied:
Good point. I will follow your suggestion to implement further extension.

joboog replied:
I would also suggest to stick to the Phreeqc flag names: `mineral` --> `name`.

renchao-lu replied:
> I would also suggest to stick to the Phreeqc flag names: `mineral` --> `name`.

Changed.

joboog commented on ChemistryLib/PhreeqcIO.cpp:
```diff
@@ -301,6 +323,41 @@ std::ostream& operator<<(std::ostream& os, PhreeqcIO const& phreeqc_io)
+                        ? chemical_system_id + 1
+                        : num_chemical_systems + chemical_system_id + 1;
+                os << "-equilibrate with solution " << aqueous_solution_id << "\n";
+                os << "-sites_units DENSITY" << "\n";
```
if you add a variable for the absolute amount of surface *sites*, defining the *-sites_units* here will be optional.
You need to add it only in case you define a surface using *site_density* in `CreateSurface.cpp`.

renchao-lu replied:
I prefer to keep the implementation as it is in this pull request, and set up another pull request  for this extension. @joboog

joboog replied:
ok.

joboog commented on ChemistryLib/PhreeqcIO.cpp:
```diff
@@ -301,6 +323,41 @@ std::ostream& operator<<(std::ostream& os, PhreeqcIO const& phreeqc_io)
+            os << "USE surface " << chemical_system_id + 1 << "\n";
+            os << "SAVE solution " << chemical_system_id + 1 << "\n";
+        }
+
```
Here you define the Phreeqc inputfile keyword blocs and call them later.
You could shorten this by:

* define `aqueous_solution`
* `if (dump)` define `aqueous_solution_prev`
* `USE solution none` 
* `END`
* `USE solution chemical_system_id + 1`
* `if (!equilibrium_phases.empty())` define `equilibrium_phase`
* `if (!kinetic_reactants.empty())`define `kinetics`
* `if (!surface.empty())` define `surface chemical_system_id + 1` (equilibrate with  aqueous_solution_prev)
* `END`

renchao-lu replied:
Very helpful. Improved as suggested.

renchao-lu commented:
Remarks:
I updated the reference results of equilibrium phase benchmark, because I added an additional step for equilibrium calculation before jumping into the time loop. In order to consolidate if my modification is correct or not in terms of both physical and technical viewpoints, please take a look where I intrude in the function `void TimeLoop::initialize()`. @HBShaoUFZ @montoyav @joboog @endJunction @wenqing

endJunction commented on ProcessLib/TimeLoop.cpp:
```diff
@@ -405,6 +405,14 @@ void TimeLoop::initialize()
+    {
+        BaseLib::RunTime time_phreeqc;
+        time_phreeqc.start();
+        _chemical_system->executeInitialCalculation(_process_solutions);
```
Just an idea; Can this initialization happen in the `ComponentTransport::setInitialConditions`?

renchao-lu replied:
Would be nice. Need some time to dig deeper.

joboog replied:
For instance, if you like to transfer the `ChemistryLib` to another process, wouldn't it be more general to have to initialization as @renchao-lu implemented?
When calling from `ComponentTransport::setInitialConditions` you would have to define the function for other processes (`Richards Flow`etc) seperately, right?

montoyav replied:
Maybe, important to know 1) how much additional computational time is needed if an additional equilibrium step is included and 2) if this additional step give more accurate results or not.

joboog replied:
The step added by @renchao-lu  will just be called once during simulation start-up. It will not substantiall increase the overall computational cost.

renchao-lu replied:
Sorry for the late reply. I just come back. As Johannes pointed out, the added equilibrium calculation is executed once before time looping.

renchao-lu commented:
- Call back the benchmark for testing and await consent of project partners.
- Completed parameter documentation.

renchao-lu commented:
Cleaned up clang-tidy warnings, while there is still no room to suppress the number of warnings under the threshold...

endJunction commented on ChemistryLib/PhreeqcIO.cpp:
```diff
@@ -80,6 +92,27 @@ PhreeqcIO::PhreeqcIO(std::string const project_file_name,
+
+    if (_dump)
+    {
+        SetDumpFileOn(phreeqc_instance_id, 1);
```
comment on the magic number would be nice

renchao-lu replied:
Added.

endJunction commented on ChemistryLib/PhreeqcIO.cpp:
```diff
@@ -174,6 +209,32 @@ void PhreeqcIO::setAqueousSolutionsOrUpdateProcessSolutions(
+                  dump_file.c_str());
+    }
+
+    in.close();
```
not needed. All objects will be destroyed at closing brace/scope.

renchao-lu replied:
I prefer to keep it if harmless...

endJunction commented on ChemistryLib/PhreeqcIO.cpp:
```diff
@@ -198,9 +259,20 @@ void PhreeqcIO::writeInputsToFile(double const dt)
     os << *phreeqc_io._output << "\n";
 
+    auto const& user_punch = phreeqc_io._user_punch;
+    if (user_punch)
```
looks like `user_punch` is a pointer. If so, it would be better to store it by value. In the current implementation you have one more indirection.
Store small objects by value...

renchao-lu replied:
Fixed. I will bear this criterion in mind.

renchao-lu replied:
_user_punch is wrapped by `unique_ptr`...

endJunction commented on ChemistryLib/PhreeqcIO.cpp:
```diff
@@ -240,9 +328,29 @@ std::ostream& operator<<(std::ostream& os, PhreeqcIO const& phreeqc_io)
         os << "END" << "\n" << "\n";
     }
 
+    auto const& dump = phreeqc_io._dump;
```
same here, store by value.

renchao-lu replied:
Fixed.

renchao-lu replied:
Wrapped by `unique_ptr` as well...

endJunction commented on ChemistryLib/PhreeqcIO.cpp:
```diff
@@ -410,6 +527,18 @@ std::istream& operator>>(std::istream& in, PhreeqcIO& phreeqc_io)
+                    auto& secondary_variable = BaseLib::findElementOrError(
+                        secondary_variables.begin(), secondary_variables.end(),
+                        compare_by_name,
+                        "Could not find secondary variable'" + item_name +
```
```suggestion
                        "Could not find secondary variable '" + item_name +
```

renchao-lu replied:
Fixed.

endJunction commented on ChemistryLib/PhreeqcIOData/CreateKnobs.cpp:
```diff
@@ -0,0 +1,53 @@
+/**
+ * \copyright
```
`\file`

renchao-lu replied:
Fixed.

endJunction replied:
Somehow not fixed...

endJunction commented on ChemistryLib/PhreeqcIOData/CreateKnobs.cpp:
```diff
@@ -0,0 +1,53 @@
+
+    auto const step_size =
+        //! \ogs_file_param{prj__chemical_system__knobs__tolerance}
+        config->getConfigParameter<int>("step_size", 100);
```
mismatch btw. `\ogs_file_param` and the actual parameter. Should be visible in the doxygen page for QC.

renchao-lu replied:
Fixed.

endJunction commented on ChemistryLib/PhreeqcIOData/CreateOutput.cpp:
```diff
@@ -48,8 +50,16 @@ std::unique_ptr<Output> createOutput(
+    {
+        auto const& secondary_variables = user_punch->secondary_variables;
+        std::transform(secondary_variables.begin(), secondary_variables.end(),
+                       std::back_inserter(accepted_items), accepted_item);
```
call to `accepted_items.reserve(accepted_items.size() + secondary_variables.size())` would save some reallocations.

renchao-lu replied:
Added.

endJunction commented on ChemistryLib/PhreeqcIOData/Dump.cpp:
```diff
@@ -0,0 +1,62 @@
+    os << "DUMP" << "\n";
+    os << "-file " << dump_file << "\n";
+    os << "-append false" << "\n";
+    os << "-solution 1-" << std::to_string(num_chemical_systems) << "\n";
```
```suggestion
    os << "-solution 1-" << num_chemical_systems << "\n";
```

renchao-lu replied:
Fixed. Too stupid... - =!

endJunction commented on ChemistryLib/PhreeqcIOData/Dump.cpp:
```diff
@@ -0,0 +1,62 @@
+
+        if (line.find("-gammas") != std::string::npos)
+        {
+            aqueous_solutions_prev.push_back(std::move(aqueous_solution_prev));
```
After the move (*i.e.* in the second iteration) `aqueous_solution_prev` must not be used...

renchao-lu replied:
Variable naming by mistake... Fixed anyway.

endJunction commented on ChemistryLib/PhreeqcIOData/Knobs.h:
```diff
@@ -0,0 +1,39 @@
+{
+struct Knobs
+{
+    Knobs(int max_iterations_, double relative_convergence_tolerance_,
```
I'd drop the ctor and the `const` from the members.

renchao-lu replied:
Changed.

endJunction commented on ChemistryLib/PhreeqcIOData/Surface.cpp:
```diff
@@ -0,0 +1,27 @@
+{
+std::ostream& operator<<(std::ostream& os, SurfaceSite const& surface_site)
+{
+    os << surface_site.name.c_str() << " " << surface_site.site_density << " "
```
```suggestion
    os << surface_site.name << " " << surface_site.site_density << " "
```

renchao-lu replied:
Fixed.

endJunction commented on ChemistryLib/PhreeqcIOData/UserPunch.cpp:
```diff
@@ -0,0 +1,40 @@
+    auto const& secondary_variables = user_punch.secondary_variables;
+    for (auto& secondary_variable : secondary_variables)
+    {
+        os << secondary_variable.name.c_str() << " ";
```
```suggestion
        os << secondary_variable.name << " ";
```

renchao-lu replied:
Fixed.

endJunction commented on ChemistryLib/PhreeqcIOData/UserPunch.cpp:
```diff
@@ -0,0 +1,40 @@
+    int line_number = 1;
+    for (auto const& statement : user_punch.statements)
+    {
+        os << line_number << " " << statement.c_str() << "\n";
```
```suggestion
        os << line_number << " " << statement << "\n";
```

renchao-lu replied:
Fixed.

endJunction commented on ChemistryLib/PhreeqcIOData/UserPunch.h:
```diff
@@ -0,0 +1,56 @@
+{
+struct SecondaryVariable
+{
+    SecondaryVariable(std::string const name_,
```
Either `const&` or value (or r-value). Be consistent.

renchao-lu replied:
Used r-value.

endJunction commented on ChemistryLib/PhreeqcIOData/UserPunch.h:
```diff
@@ -0,0 +1,56 @@
+{
+    UserPunch(std::vector<SecondaryVariable>&& secondary_variables_,
+              std::vector<std::string>&& statements_)
+        : secondary_variables(secondary_variables_), statements(statements_)
```
There are unnecessary copies here.

renchao-lu replied:
?

endJunction replied:
```suggestion
        : secondary_variables(std::move(secondary_variables_)), statements(std::move(statements_))
```

renchao-lu replied:
Fixed.

renchao-lu commented on ChemistryLib/PhreeqcIOData/CreateKnobs.cpp:
```diff
@@ -0,0 +1,53 @@
+
+    auto const max_iterations =
+        //! \ogs_file_param{prj__chemical_system__knobs__max_iterations}
+        config->getConfigParameter<int>("max_iter", 100);
```
100% sure. Do you mean to treat the numerical settings, e.g., maximum number of iterations, relative convergence tolerance, step size, and etc, as the mandatory field in the `.prj` file?

endJunction replied:
Yes

renchao-lu replied:
Added.

endJunction commented on ChemistryLib/PhreeqcIOData/CreateKnobs.cpp:
```diff
@@ -0,0 +1,45 @@
+{
+    auto const max_iterations =
+        //! \ogs_file_param{prj__chemical_system__knobs__max_iterations}
+        config.getConfigParameter<int>("max_iter", 100);
```
Didn't you change that part to be mandatory? At least I've seen the project files being updated...

renchao-lu replied:
Dropped the default values...

endJunction commented on ChemistryLib/PhreeqcIOData/Dump.cpp:
```diff
@@ -0,0 +1,63 @@
+
+        if (line.find("-gammas") != std::string::npos)
+        {
+            aqueous_solutions_prev.push_back(std::move(aqueous_solution_prev_));
```
You cannot use a moved variable. Move `aqueous_solution_prev_` inside the loop.

renchao-lu replied:
Dropped std::move.

endJunction commented on ChemistryLib/PhreeqcIOData/Dump.h:
```diff
@@ -0,0 +1,36 @@
+{
+struct Dump
+{
+    explicit Dump(std::string project_file_name)
```
const &

renchao-lu replied:
Added.

endJunction commented on Documentation/ProjectFile/prj/chemical_system/knobs/t_max_iterations.md:
```diff
@@ -0,0 +1 @@
+an optional tag for setting maximum number of iteration. The default value is 100.
```
no default

renchao-lu replied:
Modified docu.

endJunction commented on ChemistryLib/PhreeqcIOData/Surface.h:
```diff
@@ -0,0 +1,47 @@
+{
+struct SurfaceSite
+{
+    SurfaceSite(std::string name_,
```
minor: I'd drop the ctor.

endJunction commented on ChemistryLib/PhreeqcIOData/UserPunch.h:
```diff
@@ -0,0 +1,58 @@
+{
+struct SecondaryVariable
+{
+    SecondaryVariable(std::string name_,
```
minor: Also this one can be dropped, as well as the `UserPunch`...

endJunction commented on Documentation/ProjectFile/prj/chemical_system/t_use_high_precision.md:
```diff
@@ -0,0 +1 @@
+Optional tag. Th default setting is to use high precision. Convergence criterion is thus set to 1e-12. When this tag is explicitly set to false, convergence criterion will be reduced to 1e-8.
```
```suggestion
Optional tag. The default setting is to use high precision. Convergence criterion is then set to 1e-12. If this tag is explicitly set to false, convergence criterion will be reduced to 1e-8.
```

endJunction commented on Documentation/ProjectFile/prj/chemical_system/user_punch/i_user_punch.md:
```diff
@@ -0,0 +1 @@
+contain a number of statements for calculating custom secondary variables.
```
```suggestion
Contains a number of statements for calculating custom secondary variables.
```

TomFischer commented on ChemistryLib/PhreeqcIO.cpp:
```diff
@@ -219,6 +290,21 @@ std::ostream& operator<<(std::ostream& os, PhreeqcIO const& phreeqc_io)
+            auto const& aqueous_solutions_prev = dump->aqueous_solutions_prev;
+            if (!aqueous_solutions_prev.empty())
+            {
+                os << aqueous_solutions_prev[chemical_system_id] << "\n" << "\n";
```
```suggestion
                os << aqueous_solutions_prev[chemical_system_id] << "\n\n";
```

TomFischer commented on ChemistryLib/PhreeqcIO.cpp:
```diff
@@ -219,6 +290,21 @@ std::ostream& operator<<(std::ostream& os, PhreeqcIO const& phreeqc_io)
+        }
+
+        os << "USE solution none" << "\n";
+        os << "END" << "\n" << "\n";
```
```suggestion
        os << "END" << "\n\n";
```

TomFischer commented on ChemistryLib/PhreeqcIO.cpp:
```diff
@@ -240,9 +327,29 @@ std::ostream& operator<<(std::ostream& os, PhreeqcIO const& phreeqc_io)
+                os << "SAVE solution " << chemical_system_id + 1 << "\n";
+        }
+
         os << "END" << "\n" << "\n";
```
```suggestion
        os << "END" << "\n\n";
```
